### PR TITLE
[AOTAutograd] fix aot autograd incorrectly assumes ambient `no_grad` even though output `requires_grad`

### LINKED ISSUE #114338: [AOTAutograd] torch.compile under ambient `no_grad` is broken
### 🐛 Describe the bug

I'm not sure how this is possible that this behaviour is occuring on main, but it is.

```python
def fn(x):
    with torch.enable_grad():
        out = x + 1
    return out
    

opt_fn = torch.compile(fn, backend="inductor")  # also, aot_eager, aot_eager_decomp_partition, but not eager
with torch.no_grad():
    res = fn(torch.zeros(10, requires_grad=True))
    opt_res = opt_fn(torch.zeros(10, requires_grad=True))

    assert res.requires_grad == opt_res.requires_grad  # True != False
```

Traced Dynamo FX graph

```
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG] TRACED GRAPH
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]  ===== __compiled_fn_0 =====
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]  <eval_with_key>.0 class GraphModule(torch.nn.Module):
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]     def forward(self, L_x_ : torch.Tensor):
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         l_x_ = L_x_
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         # No stacktrace found for following nodes
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         _set_grad_enabled = torch._C._set_grad_enabled(True)
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         # File: /home/jonch/Desktop/sdpa.py:2169, code: out = x + 1
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         out = l_x_ + 1;  l_x_ = None
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         # No stacktrace found for following nodes
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         _set_grad_enabled_1 = torch._C._set_grad_enabled(False)
[2023-11-22 00:01:20,308] [0/0] torch._dynamo.output_graph.__graph_code: [DEBUG]         return (out,)

```
### Versions

main

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/114338

This is a hotfix, but it seems there is a nest of issues here potentially.

Not fixed:
1. Sister issue (not fixed): https://github.com/pytorch/pytorch/issues/114344
2. Note that there may also be another subtle issue in https://github.com/pytorch/pytorch/blob/4e4a6ad6ecd71a1aefde3992ecf7f77e37d2e264/torch/_functorch/aot_autograd.py#L4270
I suspect we need to change and to or / else rely on collected output metas to determine if any output requires_grad. This would require additional testing/exploration. (or maybe it is fine? It could just be to avoid nested grad)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -3205,7 +3205,7 @@ def runtime_wrapper(*args):
             for idx in indices_of_inps_to_detach:
                 if isinstance(args_[idx], torch.Tensor):
                     args_[idx] = args_[idx].detach()
-            with torch.autograd._force_original_view_tracking(True):
+            with torch.enable_grad(), torch.autograd._force_original_view_tracking(True):
```
Comment (jon-chuang) [inline torch/_functorch/aot_autograd.py:3208]: I don't completely understand why this seems to fix things, but it does.

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -4458,7 +4458,7 @@ def convert(idx, x):
 
         needs_autograd = (
             any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
-            and torch.is_grad_enabled()
+            or torch.is_grad_enabled()
```
Comment (jon-chuang) [inline torch/_functorch/aot_autograd.py:4461]: This is definitely a good idea. We already check later on whether any output requires_grad, and if not, fallback on `needs_autograd=False`

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_functorch/aot_autograd.py:3214]: I've added even more changes lessening the dependence on ambient `no_grad` and emphasizing more whether any output `requires_grad`. `needs_autograd` is controlled by the `AOTConfig.allow_joint_graph` which is by default `True`. 

In the export setting, one is required to be explicit about whether the joint graph is traced, as there are restrictions on this joint graph.

In the compile setting, we will create backwards function as long as there is an output with `requires_grad`

**Comment by jon-chuang** [comment]
There were some maybe related changes on main, let's see if the tests pass without this PR.

**Comment by ezyang** [comment]
Bumping this

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -4456,9 +4462,9 @@ def convert(idx, x):
 
         fake_flat_args = process_inputs(flat_args)
 
-        needs_autograd = (
+        needs_autograd = aot_config.allow_joint_graph and (
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:4465]: Hey sorry for dropping this @jon-chuang.

Hmm, I don't love the idea of a config - ideally we should just detect whether we need a training vs. inference graph in all cases automatically.

Now that we are basically detecting "do we need a training graph" in two steps, the first step here is basically just a heuristic (if we do need a training graph then this should definitely be set to True, but if it turn out that we don't need a training graph, because all the graph outputs don't require grad, then it's ok to pessimistically set this to true, and set it to false lower down).

So in theory we could just set this to `True` here, although one downside right now is that getting it wrong means that we need to re-run `run_functionalized_fw_and_collect_metadata` in the subclass path, which will hurt compile perf.

Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:4465]: Instead of the new config, can you just make the condition this?
```
needs_autograd = any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
# forget about looking at `is_grad_enabled`
```

That should fix this case too (I had the same idea [here](https://github.com/pytorch/pytorch/pull/115294), but happy to land you PR), since we won't incorrect label the test you've written as an inference graph.

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:3214]: yep, the `enable_grad` is definitely more correct - if we knew at trace-time that we needed a training graph, then we definitely need the ambient grad state to be turned on at runtime.

If it turns out that some of our outputs should rquire grad and some should not, that should be handled by the fact we're using `ctx.mark_non_differentiable()` to set some outputs as not requiring grad (based on what we detected at tracetime), [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L507)
